File: frontend/src/api/system-settings.ts
import { apiClient } from "./client"
import { BASE_URL } from "./constants"

export const SYSTEM_SETTING_CURRENT_SEMESTER = "current_semester"
export const SYSTEM_SETTING_AUTH_EMAIL_DOMAIN = "auth.email_domain"

export type SystemSettingType =
  | "string"
  | "int"
  | "bool"
  | "float"
  | "duration"
  | "string_list"

export interface SystemSettingDTO {
  key: string
  value: string
  default_value: string
  type: SystemSettingType
  group: string
  label: string
  description: string
  public: boolean
  secret: boolean
  requires_restart: boolean
}

export interface UpdateSystemSettingCommand {
  value: string
}

export function listSystemSettings(): Promise<SystemSettingDTO[]> {
  return apiClient(`${BASE_URL}/system-settings`)
}

export function listAdminSystemSettings(): Promise<SystemSettingDTO[]> {
  return apiClient(`${BASE_URL}/admin/system-settings`)
}

export function updateSystemSetting(
  key: string,
  cmd: UpdateSystemSettingCommand
): Promise<SystemSettingDTO> {
  return apiClient(
    `${BASE_URL}/admin/system-settings/${encodeURIComponent(key)}`,
    {
      method: "PUT",
      body: JSON.stringify(cmd),
    }
  )
}

File: backend/internal/application/system_settings.go
package application

import "jcourse/internal/domain/setting"

type SystemSettingDTO struct {
	Key             string `json:"key"`
	Value           string `json:"value"`
	DefaultValue    string `json:"default_value"`
	Type            string `json:"type"`
	Group           string `json:"group"`
	Label           string `json:"label"`
	Description     string `json:"description"`
	Public          bool   `json:"public"`
	Secret          bool   `json:"secret"`
	RequiresRestart bool   `json:"requires_restart"`
}

type UpdateSystemSettingCommand struct {
	Value string `json:"value"`
}

func newSystemSettingDTO(item setting.EffectiveSystemSetting) SystemSettingDTO {
	def := item.Definition
	return SystemSettingDTO{
		Key:             def.Key,
		Value:           item.Value,
		DefaultValue:    def.DefaultValue,
		Type:            string(def.Type),
		Group:           def.Group,
		Label:           def.Label,
		Description:     def.Description,
		Public:          def.Public,
		Secret:          def.Secret,
		RequiresRestart: def.RequiresRestart,
	}
}
